mathesar_ui/src/components/sheet/SheetClipboardHandler.ts

```diff
@@ -0,0 +1,143 @@
+import { get } from 'svelte/store';
+
+import { ImmutableSet, type MakeToast } from '@mathesar-component-library';
+import SheetSelection, {
+  isCellSelected,
+} from '@mathesar/components/sheet/SheetSelection';
+import type { ClipboardHandler } from '@mathesar/stores/clipboard';
+import type {
+  ProcessedColumn,
+  RecordRow,
+  RecordSummariesForSheet,
+} from '@mathesar/stores/table-data';
+import type { QueryRow } from '@mathesar/systems/data-explorer/QueryRunner';
+import type { ProcessedQueryOutputColumn } from '@mathesar/systems/data-explorer/utils';
+import type { ReadableMapLike } from '@mathesar/typeUtils';
+import { labeledCount } from '@mathesar/utils/languageUtils';
+
+const MIME_PLAIN_TEXT = 'text/plain';
+const MIME_MATHESAR_SHEET_CLIPBOARD =
+  'application/x-vnd.mathesar-sheet-clipboard';
+
+type CopyingStrategy = 'raw' | 'formatted';
+
+/** Keys are row ids, values are records */
+type IndexedRecords = Map<number, Record<string, unknown>>;
+
+function getCellText<
+  Column extends ProcessedQueryOutputColumn | ProcessedColumn,
+>(
+  indexedRecords: IndexedRecords,
+  columnsMap: ReadableMapLike<Column['id'], Column>,
+  rowId: number,
+  columnId: Column['id'],
+  strategy: CopyingStrategy,
+  recordSummaries: RecordSummariesForSheet,
+): string {
+  const record = indexedRecords.get(rowId);
+  if (!record) {
+    return '';
+  }
+  const rawCellValue: unknown = record[String(columnId)];
+  if (rawCellValue === undefined || rawCellValue === null) {
+    return '';
```

Comment:
We should probably document `null` values would be copied as `''` in the clipboard specs.